tests/test_argparcel.py

```diff
@@ -401,3 +401,147 @@ class _Moo:
 
     with pytest.raises(ValueError, match="Need exactly one type of choice"):
         assert _parse(_Moo, "--x")
+
+
+def test_tuple_unannotated() -> None:
+    # We can only parse a tuple if the element types are specified
+    @dataclasses.dataclass(kw_only=True, frozen=True, slots=True)
+    class _Moo:
+        x: tuple  # pyright: ignore [reportMissingTypeArgument]
+
+    with pytest.raises(ValueError, match="`tuple` must be subscripted"):
+        _parse(_Moo, "--x 1")
+
+
+def test_tuple_empty() -> None:
+    @dataclasses.dataclass(kw_only=True, frozen=True, slots=True)
+    class _Moo:
+        x: tuple[()]
+
+    with pytest.raises(ValueError, match="Empty tuples not supported"):
+        _parse(_Moo, "--x")
+
+
+def test_tuple_int1() -> None:
+    @dataclasses.dataclass(kw_only=True, frozen=True, slots=True)
+    class _Moo:
+        x: tuple[int]
+
+    assert _parse(_Moo, "--x 1").x == (1,)
+
+    with pytest.raises(argparse.ArgumentError, match="expected 1 argument"):
+        _parse(_Moo, "--x")
+
+    with pytest.raises(argparse.ArgumentError, match="unrecognized arguments: 2"):
+        _parse(_Moo, "--x 1 2")
+
+
+def test_tuple_int1_none() -> None:
+    @dataclasses.dataclass(kw_only=True, frozen=True, slots=True)
+    class _Moo:
+        x: tuple[int] | None = None
+
+    assert _parse(_Moo, "").x is None
+    assert _parse(_Moo, "--x 1").x == (1,)
+
+    with pytest.raises(argparse.ArgumentError, match="expected 1 argument"):
+        _parse(_Moo, "--x")
+
+    with pytest.raises(argparse.ArgumentError, match="unrecognized arguments: 2"):
+        _parse(_Moo, "--x 1 2")
+
+
+def test_tuple_int2() -> None:
+    @dataclasses.dataclass(kw_only=True, frozen=True, slots=True)
+    class _Moo:
+        x: tuple[int, int]
+
+    assert _parse(_Moo, "--x 1 2").x == (1, 2)
+
+    with pytest.raises(argparse.ArgumentError, match="expected 2 arguments"):
+        _parse(_Moo, "--x")
```

Comment:
Is it an implementation detail that `tuple[int, ...]` appears inhomogeneous, because the `...` is different from `int`? Perhaps this could be clearer with a special-cased error message that only fixed-length tuples are supported. And optionally to recommend `list[int]` for flexible length?

Reply:
https://github.com/tpgillam/argparcel/issues/36